synapseclient/multipart_upload.py

```diff
@@ -0,0 +1,321 @@
+            printTransferProgress(completed.value, fileSize, prefix='Uploading', postfix=filename)
+    except IOError as ex1:
+        print(ex1)
+        print("Encountered an exception: %s. Retrying..." % type(ex1))
+
+
+def _multipart_upload(syn, filename, contentType, get_chunk_function, md5, fileSize, partSize, retries=7, url_batch_size=6):
+    """
+    Multipart Upload.
+
+    :param syn: a Synapse object
+    :param filename: a string containing the base filename
+    :param contentType: contentType_
+    :param get_chunk_function: a function that takes a part number and size
+                               and returns the bytes of that chunk of the file
+    :param md5: the part's MD5 as hex.
+    :param fileSize: total number of bytes
+    :param partSize: number of bytes per part. Minimum 5MB.
+    :param retries: number of times to retry upload
+    :param url_batch_size: number of signed URLs to request at once
+
+    :return: a MultipartUploadStatus_ object
+
+    .. MultipartUploadStatus: http://rest.synapse.org/org/sagebionetworks/repo/model/file/MultipartUploadStatus.html
+    .. contentType: https://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html#sec14.17
+    """
+    from multiprocessing import Value
+
+    if partSize is None:
+        partSize = calculate_part_size(fileSize, MAX_NUMBER_OF_PARTS)
+    if partSize < MIN_PART_SIZE:
+        raise ValueError('Minimum part size is %d MB.' % (MIN_PART_SIZE/MB))
+
+    ## make upload_chunk a function of 4 parameters:
+    ##    partNumber, url, completed, status
+    upload_chunk = partial(_upload_chunk, syn=syn,
+                                          filename=filename,
+                                          get_chunk_function=get_chunk_function,
+                                          fileSize=fileSize,
+                                          partSize=partSize)
+
+    for i in range(retries):
+        status = _start_multipart_upload(syn, filename, md5, fileSize, partSize, contentType)
+
+        completed = Value('d', min(count_completed_parts(status.partsState) * partSize, fileSize))
+        printTransferProgress(completed.value, fileSize, prefix='Uploading', postfix=filename)
+
+        ## for each batch of parts to upload, get presigned URLs and attempt upload
+        for parts_to_upload in partition(url_batch_size, find_parts_to_download(status.partsState)):
+            presigned_url_batch = _get_presigned_urls(syn, status.uploadId, parts_to_upload)
+            for part in presigned_url_batch.partPresignedUrls:
+                upload_chunk(partNumber=part["partNumber"],
+                             url=part["uploadPresignedUrl"],
+                             completed=completed,
+                             status=status)
+
+        ## Are we done, yet?
+        if completed.value >= fileSize:
+            status = _complete_multipart_upload(syn, status.uploadId)
+            if status.state == "COMPLETED":
+                break
+
```

Comment:
could complete_multipart_upload always return a status object and not throw an exception?